Add guide (#1)
hazelcast-guides/ec2-cluster: doc/modules/ROOT/pages/index.adoc

of instances is 2. Also, enable `Auto-assign Public IP` to establish SSH connections later on.

image::ss5.3-create-instance.png[]

* Next, add a unique tag to the instances. This is optional but recommended if your AWS account has many running
instances associated with:

image::ss5.4-create-instance.png[]

* Finally, select the security group you created above:

image::ss5.5-create-instance.png[]

As the last step, select your https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html[key pair]
for the instances and that's it. You can launch instances now.

== 6) Create a Hazelcast Cluster

* Now that you have 2 instances running with the same IAM role, let's connect to each of them via SSH:

[source, shell]
----
ssh -i "<your-key-pair>.pem" ec2-user@<instance-public-ip>
----

[NOTE]
====
If you encounter any problem regarding the SSH connection, see
https://aws.amazon.com/tr/premiumsupport/knowledge-center/ec2-linux-ssh-troubleshooting/[SSH troubleshooting page]
in AWS documentation.
====

* After SSH connection is established, install Hazelcast CLI to the instances:

[source, shell]
----

> wget https://bintray.com/hazelcast/rpm/rpm -O bintray-hazelcast-rpm.repo && \
> sudo mv bintray-hazelcast-rpm.repo /etc/yum.repos.d/ && \
> sudo yum install hazelcast

Reviewer: ```suggestion
wget https://bintray.com/hazelcast/rpm/rpm -O bintray-hazelcast-rpm.repo
sudo mv bintray-hazelcast-rpm.repo /etc/yum.repos.d/
sudo yum install hazelcast
```
That way should be fine as well.

:github-address: https://github.com/hazelcast-guides/ec2-cluster
:templates-url: templates:ROOT:page$/
:aws-plugin-url: https://github.com/hazelcast/hazelcast-aws

= Deploy Hazelcast Cluster on EC2

This guide will get you started to create a Hazelcast cluster on EC2.

== What You’ll Learn

In this guide, you will create two EC2 instances with Hazelcast members and see them connecting each other and
forming a cluster using {aws-plugin-url}[Hazelcast AWS plugin].

[NOTE]
====

> Steps 1 to 4 cover the creation and the configuration of VPC, internet gateway, route table, subnet and IAM rule.

Reviewer: ```suggestion
Steps 1 to 4 cover the creation and the configuration of VPC, internet gateway, route table, subnet and IAM role.
```

If you already have these configured you can skip them. However, make sure that you have these properties before
creating instances:

> * [x] Have an IAM rule with `ec2:DescribeInstances` permission

Reviewer: ```suggestion
* [x] Have an IAM role with `ec2:DescribeInstances` permission
```

> * [x] Have a security group with ports 5701-5708 (or whatever to be used by Hazelcast) opened

Reviewer: ```suggestion
* [x] Have a security group with at least port 5701 open
```
I think we can suggest opening just 5701.

* [x] Able to establish SSH connections to instances
====

> [NOTE]

Reviewer: Reading this, I'm not sure what is not secure and why we can't use it for the production environments. I'd suggest to remove this note and maybe add some others inside the steps that are not safe.

Author: I decided to add this section after seeing aws warning when allowing all source IPs for inbound rules. But I agree with you to add this just before the related section - if we want to.

====
The configurations below are not appropriate and not secure for production environments and should be used
only in test environments.
====

== 1) Configuring VPC

Let's start with creating a VPC on which our instances will run:

=== 1.1) Create a VPC

Under `VPC > Your VPCs` section of AWS console, create a new VPC and give it a name you can remember in the next steps.

image::ss1-create-vpc.png[]

> ==== 1.1.a) Create an Internet Gateway

Reviewer: Nit: this enumeration style `1.1.a)` is slightly weird, how about keeping the numbers only and changing it to `1.1.1)`?

Now create an internet gateway to be attached to your VPC. You will need this gateway to establish SSH connections to
instances:

image::ss1.1.1-create-igw.png[]

==== 1.1.b) Attach the Internet Gateway

After creating an internet gateway, attach it to your VPC on `VPC > Internet gateways > Attach to VPC` section.
Choose the VPC you just created among the available ones:

image::ss1.1.2-attach-igw.png[]

=== 1.2) Add Default Route

Now add default route to the route table. This step will make you able to connect to instances via SSH.

Under the VPC description, navigate to route table shown in the orange box:

image::ss1.2.1-nav-to-routes.png[]

> Under `Routes` tab, ypu need the default one (`0.0.0.0/0`) listed. `Edit routes` to add this one:

Reviewer: ```suggestion
Under `Routes` tab, you need the default one (`0.0.0.0/0`) listed. `Edit routes` to add this one:
```

image::ss1.2.2-list-routes.png[]

As target, pick the internet gateway you created in section `1.1.a` and save:

image::ss1.2.3-add-route.png[]

== 2) Create a Subnet in the VPC

Let's continue with creating a subnet in the VPC. Under `VPC > Subnets` section, choose `Create Subnet`. Pick the proper
VPC and give the subnet a recognizable name:

image::ss2-create-subnet.png[]

== 3) Create an IAM Role

The EC2 instances we will create need your IAM role to have `ec2:DescribeInstances` permission. This way, Hazelcast
members will be able to fetch other instance IPs and connect them dynamically. If you already have an IAM role, check
for the permission. Otherwise, create a new one under `IAM > Roles` section with the mentioned permission. For instance,

> `AmazonEC2ReadOnlyAccess` policy contains `DescribeInstances` permission and is enough to complete this guide:

Reviewer: ```suggestion
The EC2 instances we will create need the IAM role to have `ec2:DescribeInstances` permission. This way, Hazelcast
members are able to fetch other instance IPs and connect them dynamically. If you already have an IAM role, check
the permissions. Otherwise, create a new one under the `IAM > Roles` section. For instance,
`AmazonEC2ReadOnlyAccess` policy contains `DescribeInstances` permission and is enough to complete this guide:
```

> 

Reviewer: How about adding one more screenshot with how to create an IAM role? Like what to select here.
![image](https://user-images.githubusercontent.com/2834997/102332015-d6b32200-3f8b-11eb-8c98-dde379f42fef.png)

image::ss3-create-iam-with-permission.png[]

== 4) Create a Security Group

> As the last step, under `VPC > Security Groups` create a security group in your VPC with the proper inbound rules

Reviewer: I'd actually suggest opening only `5701`. And mention that if you plan to run more Hazelcast members on one EC2 Instance, then you should open more ports.

for Hazelcast. Although we need only 5701 in this guide, allow the port range 5701-5708 among inbound rules as it's
the default range for Hazelcast AWS discovery plugin. Do not forget to allow SSH port also:

image::ss4-create-security-group.png[]

== 5) Create EC2 Instances

Let's start creating our instances via `LaunchInstanceWizard` under `EC2 > Launch Instances` on AWS Console.

* Choose an Amazon Machine Image (AMI). `Amazon Linux` is used in this guide:

image::ss5.1-create-instance.png[]

* Choose an instance type:

image::ss5.2-create-instance.png[]

* Now configure instance details with the *_VPC_, _subnet_ and _IAM roles_* you created above. Notice that the number
of instances is 2. Also, enable `Auto-assign Public IP` to establish SSH connections later on.

image::ss5.3-create-instance.png[]

* Next, add a unique tag to the instances. This is optional but recommended if your AWS account has many running
instances associated with. We will configure Hazelcast to filter discovered EC2 instances based on this tag:

image::ss5.4-create-instance.png[]

> * Finally, select the security group you created above. You also can use different security groups for different

Reviewer: I'd skip the information that you can use different security groups. It's not relevant in my opinion.

instances as long as you do not set `security-group-name` property in Hazelcast AWS configuration (more on configuration
details in the next section):

image::ss5.5-create-instance.png[]

As the last step, select your https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html[key pair]
for the instances and that's it. You can launch instances now.

== 6) Create a Hazelcast Cluster

* Now that you have 2 instances running with the same IAM role, let's connect to each of them via SSH:

[source, shell]
----
ssh -i "<your-key-pair>.pem" ec2-user@<instance-public-ip>
----

[NOTE]
====
If you encounter any problem regarding the SSH connection, see
https://aws.amazon.com/tr/premiumsupport/knowledge-center/ec2-linux-ssh-troubleshooting/[SSH troubleshooting page]
in AWS documentation.
====

* After SSH connection is established, install JDK8 (or any newer version) to the instances:

[source, shell]
----

> sudo yum -y update && sudo yum install java-1.8.0-openjdk

Reviewer: I'd suggest using [Hazelcast CLI](https://github.com/hazelcast/hazelcast-command-line). So the next code could look like this:

```
wget https://bintray.com/hazelcast/rpm/rpm -O bintray-hazelcast-rpm.repo
sudo mv bintray-hazelcast-rpm.repo /etc/yum.repos.d/
sudo yum install hazelcast
hz start
```

So you can show that it just works and then give an example with filtering by tag.

----

* Verify that it is installed:

[source, shell]
----
java -version
----

* Now, let's install Hazelcast JAR. Do not forget to set *HZ_VERSION* variable or change the URL with the version you use:

[source, shell]
----
JAR_URL="https://repo1.maven.org/maven2/com/hazelcast/hazelcast-all/${HZ_VERSION}/hazelcast-all-${HZ_VERSION}.jar"
curl -sf -O -L $JAR_URL
----

[NOTE]
====
To make instances find each other and form a cluster, {aws-plugin-url}[Hazelcast AWS plugin]
must be used. This is included in `hazelcast-all` and you are all set if you downloaded from the URL above. Otherwise,
you need to include `hazelcast-aws` explicitly.
====

> * Create a Hazelcast configuration now. A minimal configuration with AWS discovery enabled will look like the yaml

Reviewer: How about first just starting `hz start` and showing that it works. And then adding this filtering by tag. Also maybe better to try to use config override than providing the whole configuration?

:github-address: https://github.com/hazelcast-guides/ec2-cluster
:templates-url: templates:ROOT:page$/
:aws-plugin-url: https://github.com/hazelcast/hazelcast-aws
:tab-header-console: AWS Console
:tab-header-cli: AWS CLI

= Deploy Hazelcast Cluster on AWS EC2

This guide will get you started to create a Hazelcast cluster on EC2.

== What You’ll Learn

In this guide, you will create two EC2 instances with Hazelcast members and see them connecting each other and

> forming a cluster using {aws-plugin-url}[Hazelcast AWS plugin].

Reviewer 2: ```suggestion
forming a cluster using by {aws-plugin-url}[Hazelcast AWS plugin].
```

[NOTE]
====
Steps 1 to 4 cover the creation and the configuration of VPC, internet gateway, route table, subnet and IAM role.
If you already have these configured you can skip them. However, make sure that you have these properties before
creating instances:

* [x] Have an IAM role with `ec2:DescribeInstances` permission
* [x] Have a security group with at least port 5701 open
* [x] Able to establish SSH connections to instances
====

> == 1) Create and Configure a VPC

Reviewer 2: Using `.` instead of `)` is more stylish. WDYT? 

1. Create and Configure a VPC
1.1. Create a VPC

  OwnerId: '**************'
  Tags: []
----

Then add a name tag to make it distinguishable:
[source, shell]
----
$ aws ec2 create-tags --resources igw-************** \
    --tags Key=Name,Value=hazelcast-guide
----
--
====

==== 1.1.2) Attach the Internet Gateway

After creating an internet gateway, attach it to your VPC.

[tabs]
====

{tab-header-console}::
+
--
Navigate to `VPC > Internet gateways > Attach to VPC` section. Choose the VPC you just created among the available ones:

image::ss1.1.2-attach-igw.png[]
--

{tab-header-CLI}::
+
--

Using their id's, attach the gateway to the VPC:

[source, shell]
----
$ aws ec2 attach-internet-gateway \
    --internet-gateway-id igw-*********** \
    --vpc-id vpc-***********
----
--
====

=== 1.2) Add Default Route

Now add the default route to the route table. This step will make you able to connect to instances via SSH.

[tabs]
====

{tab-header-console}::
+
--
Under the VPC description, navigate to route table shown in the orange box:

image::ss1.2.1-nav-to-routes.png[]

Under `Routes` tab, you need the default one (`0.0.0.0/0`) listed. `Edit routes` to add this one:

image::ss1.2.2-list-routes.png[]

> As the target, pick the internet gateway you created in section `1.1.1` and save:

Reviewer 2: I think, in-page section reference would be better for better user-experience:
https://docs.antora.org/antora/2.3/asciidoc/in-page-xref/
